radlab-ui/webapp/src/components/forms/fields/MapNestedField.tsx

```diff
@@ -0,0 +1,179 @@
+  const [isValidParentKey, setIsValidParentKey] = useState(false)
+  const [isParentKeyEmpty, setIsParentKeyEmpty] = useState(false)
+
+  //for map nested field type
+  const handleChangeParentKey = (
+    event: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    setMapNestedParentFieldKey(event.target.value)
+    const validationError = validate(event.target.value)
+    setValidErrorMessage(validationError)
+    validationError ? setIsValidParentKey(false) : setIsValidParentKey(true)
+  }
+
+  const handleChangeKey = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setMapNestedFieldKey(event.target.value)
+    const validationError = validate(event.target.value)
+    setValidErrorMessage(validationError)
+    validationError ? setIsValidKey(false) : setIsValidKey(true)
+  }
+  const handleChangeValue = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setMapNestedFieldValue(event.target.value)
+    const validationError = validate(event.target.value)
+    setValidErrorMessage(validationError)
+    validationError ? setIsValidValue(false) : setIsValidValue(true)
+  }
+
+  const onSetKeyValue =
+    (attributeItems: IObjKeyPair) =>
+    (event: React.MouseEvent<HTMLButtonElement>) => {
+      event.preventDefault()
+      if (mapNestedFieldKey && mapNestedFieldValue) {
+        const checkLastKey =
+          Object.keys(attributeItems)[Object.keys(attributeItems).length - 1]
+        //@ts-ignore
+        attributeItems[checkLastKey][mapNestedFieldKey] = mapNestedFieldValue
+        setIsParentKeyEmpty(false)
+      }
+      setMapNestedFieldKey("")
+      setMapNestedFieldValue("")
+      setIsValidKey(false)
+      setIsValidValue(false)
+    }
+
+  const onSetParentKeyValue =
+    (attributeItems: IObjKeyPair) =>
+    (event: React.MouseEvent<HTMLButtonElement>) => {
+      event.preventDefault()
+      if (mapNestedParentFieldKey) {
+        // @ts-ignore
+        attributeItems[mapNestedParentFieldKey] = new Object()
+        setIsParentKeyEmpty(true)
+      }
+      setMapNestedParentFieldKey("")
+      setMapNestedFieldKey("")
+      setMapNestedFieldValue("")
+      setIsValidParentKey(false)
+      setIsValidKey(false)
+      setIsValidValue(false)
+    }
+
+  const handleRemove =
+    (attributeItems: IObjKeyPair, attributeKey: string) =>
+    (event: React.MouseEvent<HTMLButtonElement>) => {
+      event.preventDefault()
+      delete attributeItems[attributeKey]
+      setDeleteItem(attributeKey)
+    }
+
+  return (
+    <div className="form-control" key={variable.name}>
+      <label htmlFor={variable.name}>{variable.display}</label>
+      <FieldArray name={variable.name}>
+        {(FieldArrayProps) => {
+          const { form } = FieldArrayProps
+          const { values } = form
+          return (
+            <div>
+              {Object.keys(values[variable.name]).map(
+                (map_data: any, index: number) => (
+                  <div
+                    key={index}
```

Comment:
Better to use a data value from the variable as the `key` instead of the `index`